File: groupAnagram.py
from collections import defaultdict
class Solution:
    def validAnagram(self, p, q):
        if len(p) != len(q):
            return False
        for idx in set(p):
            if p.count(idx) != q.count(idx):
                return False
        return True
    
    def groupAnagram(self, strs):
        res = defaultdict(list)
        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord("a")] += 1
                print(c, count)
                print("\n")
            res[tuple(count)].append(s)
            print(res)
        return res.values()



            
        

a = Solution()
b = a.groupAnagram(["eat","tea","tan","ate","nat","bat"])
print(b)

File: javascript/dsa/groupAnagram.js
var groupAnagrams = function (strs) {
    const dict = {};
    for (let i = 0; i < strs.length; i++) {
        const chars = Array(26).fill(0);
        for (let j = 0; j < strs[i].length; j++) {
            const charCode = strs[i][j].charCodeAt(0) - "a".charCodeAt(0);
            chars[charCode] += 1;
        }
        // Use the character count array as a key directly
        const key = chars.join(",");

        if (dict[key] === undefined) {
            dict[key] = [];
        }

        dict[key].push(strs[i]);
    }
    return Object.values(dict);
};

console.log(groupAnagrams(["eat", "tea", "tan", "ate", "nat", "bat"]));
